easy-rules-jexl/src/test/java/org/jeasy/rules/jexl/JexlActionTest.java

```diff
@@ -0,0 +1,219 @@
+        Action printAction = new JexlAction(
+                "var random = function() { System.out.println(\"Random from JEXL = \" + new('java.util.Random', 123).nextInt(10)); }; random();",
+                jexl);
+        Facts facts = new Facts();
+        facts.put("System", System.class);
+
+        // when
+        printAction.execute(facts);
+
+        // then
+        assertThat(systemOutRule.getLog()).contains("Random from JEXL = 2");
+    }
+
+    @Test
+    public void testWithBlackSanbox() throws Exception {
+        JexlSandbox sandbox = new JexlSandbox(false);
+        sandbox.white(System.class.getName()).execute("currentTimeMillis");
+        JexlEngine jexl = new JexlBuilder()
+                .sandbox(sandbox)
+                .create();
+
+        Facts facts = new Facts();
+        facts.put("result", null);
+        facts.put("System", System.class);
+        facts.put("UUID", UUID.class);
+
+        long now = System.currentTimeMillis();
+        new JexlAction("result = System.currentTimeMillis()", jexl).execute(facts);
+        Object result = facts.get("result");
+        assertThat(result).isInstanceOf(Long.class);
+        assertThat((long) result).isGreaterThanOrEqualTo(now);
+
+        try {
+            new JexlAction("result = System.nanoTime()", jexl).execute(facts);
+            fail("Exception expected");
+        } catch (Exception e) {
+            assertThat(e).isInstanceOf(JexlException.Method.class);
+        }
+
+        try {
+            new JexlAction("result = new('java.lang.Double', 10)", jexl).execute(facts);
+            fail("Exception expected");
+        } catch (Exception e) {
+            assertThat(e).isInstanceOf(JexlException.Method.class);
+        }
+
+        try {
+            new JexlAction("result = new('java.io.File', '/tmp/jexl-sandbox-test')", jexl).execute(facts);
+            fail("Exception expected");
+        } catch (Exception e) {
+            assertThat(e).isInstanceOf(JexlException.Method.class);
+        }
+
+        try {
+            new JexlAction("result = UUID.randomUUID()", jexl).execute(facts);
+            fail("Exception expected");
+        } catch (Exception e) {
+            assertThat(e).isInstanceOf(JexlException.Method.class);
+        }
+
+        try {
+            new JexlAction("result = Math.PI", jexl).execute(facts);
+            fail("Exception expected");
+        } catch (Exception e) {
+            assertThat(e).isInstanceOf(JexlException.Variable.class);
+        }
+    }
+
+    @Test
+    public void testWithWhiteSanbox() throws Exception {
+        JexlSandbox sandbox = new JexlSandbox(true);
+        sandbox.black(System.class.getName()).execute("nanoTime");
+        JexlEngine jexl = new JexlBuilder()
+                .sandbox(sandbox)
+                .create();
+
+        Facts facts = new Facts();
+        facts.put("result", null);
+        facts.put("System", System.class);
+        facts.put("UUID", UUID.class);
```

Comment:
Facts are designed to store business data. I'm not a fan of putting technical classes as facts like in here. MVEL and SpEL provide a `ParserContext` API that allows to register/import packages and functions (See example [here](https://github.com/j-easy/easy-rules/blob/5d5b5333f9d8d4305b28578490e13b3c56036356/easy-rules-mvel/src/test/java/org/jeasy/rules/mvel/MVELActionTest.java#L93)). Is there something similar in JEXL?

My big concern here is that those tests serve as documentation as well, and in my experience, people will copy/paste and misuse facts to store any kind of non business facts.

Reply:
> Is there something similar in JEXL?

It looks like Jexk `namespaces` are the closed concept to `ParserContext` in Mvel/SpEL. Here is a quick example:

```java
	@Test
	public void testJexlNamespaces() {
		Map<String, Object> namspaces = new HashMap<>();
		namspaces.put("greeter", new Greeter());
		JexlEngine jexlEngine = new JexlBuilder()
				.namespaces(namspaces)
				.create();
		MapContext mapContext = new MapContext();
		mapContext.set("name", "foo");
		JexlScript jexlScript = jexlEngine.createScript("greeter:sayHello(name)");
		jexlScript.execute(mapContext);// prints "hello foo"
	}

	public static class Greeter {
		public void sayHello(String name) {
			System.out.println("hello " + name);
		}
	}
```